core/src/main/java/io/methvin/watcher/OnTimeoutListener.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.methvin.watcher;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+public class OnTimeoutListener implements DirectoryChangeListener {
+
+  private final int timeoutMillis;
+  private final AtomicReference<ScheduledFuture> currentTaskRef = new AtomicReference<>();
+  private ScheduledExecutorService service;
+  private Consumer<Integer> consumer;
+
+  public OnTimeoutListener(int timeoutMillis) {
+    this.timeoutMillis = timeoutMillis;
+
+    if (timeoutMillis >= 0) {
+      service = Executors.newSingleThreadScheduledExecutor();
+    }
+  }
+
+  @Override
+  public void onEvent(DirectoryChangeEvent event) {
+    ScheduledFuture taskToCancel = currentTaskRef.getAndSet(null);
+    if (taskToCancel != null) {
+      taskToCancel.cancel(false);
+    }
+  }
+
+  @Override
+  public void onIdle(int count) {
+    if (timeoutMillis >= 0) {
+      currentTaskRef.getAndUpdate(
+          oldTask -> {
+            if (oldTask != null) {
+              oldTask.cancel(false);
+            }
+            return service.schedule(
+                () -> {
+                  if (consumer != null) {
+                    consumer.accept(count);
+                  }
+                },
+                timeoutMillis,
+                TimeUnit.MILLISECONDS);
+          });
+    }
+  }
+
+  public OnTimeoutListener onTimeout(Consumer<Integer> consumer) {
+    this.consumer = consumer;
```

Comment:
I'd prefer to keep the API immutable if possible, and as-is this API doesn't make it obvious that this will replace the existing consumer.

I would suggest we pass a single `Consumer onIdleTimeout` in the constructor, or in a builder.